# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webcodecs/VideoEncoder.cpp
# Commit: b1577155be1c
# Full Hash: b1577155be1c9e9070c91a969c0cdcfaab9c8936
# Author: Chun-Min Chang <chun.m.chang@gmail.com>
# Date: 2023-12-23 09:52:03
# Regressor Bug: 1749047
# File Overlap Count: 1
# Description:
#   Bug 1749047 - harzar fix. CLOSED TREE
#   
#   
#   This patch uses RootedDictionary around `EncodedVideoChunkMetadata`,
#   `VideoDecoderConfig` and `VideoColorSpaceInit` in `OutputEncodedData` to
# ==============================================================================

diff -r ecf1c8556163 -r b1577155be1c dom/media/webcodecs/VideoEncoder.cpp
--- a/dom/media/webcodecs/VideoEncoder.cpp	Sat Dec 23 10:32:48 2023 +0200
+++ b/dom/media/webcodecs/VideoEncoder.cpp	Sat Dec 23 08:17:11 2023 +0000
@@ -586,43 +586,29 @@
   return encodedVideoChunk;
 }
 
-VideoDecoderConfig VideoEncoder::EncoderConfigToDecoderConfig(
+VideoDecoderConfigInternal VideoEncoder::EncoderConfigToDecoderConfig(
     nsIGlobalObject* aGlobal, const RefPtr<MediaRawData>& aRawData,
     const VideoEncoderConfigInternal& mOutputConfig) const {
-  VideoDecoderConfig decoderConfig;
-  decoderConfig.mCodec = mOutputConfig.mCodec;
-  decoderConfig.mCodedWidth.Construct(mOutputConfig.mWidth);
-  decoderConfig.mCodedHeight.Construct(mOutputConfig.mHeight);
-  if (mOutputConfig.mDisplayWidth.isSome()) {
-    MOZ_ASSERT(mOutputConfig.mDisplayHeight.isSome());
-    decoderConfig.mDisplayAspectWidth.Construct(
-        mOutputConfig.mDisplayWidth.value());
-    decoderConfig.mDisplayAspectHeight.Construct(
-        mOutputConfig.mDisplayHeight.value());
-  }
+  // Colorspace is mandatory when outputing a decoder config after encode
+  VideoColorSpaceInternal init;
+  init.mFullRange.emplace(false);
+  init.mMatrix.emplace(VideoMatrixCoefficients::Bt709);
+  init.mPrimaries.emplace(VideoColorPrimaries::Bt709);
+  init.mTransfer.emplace(VideoTransferCharacteristics::Bt709);
 
-  // Colorspace is mandatory when outputing a decoder config after encode
-  VideoColorSpaceInit init;
-  init.mFullRange = false;
-  init.mMatrix = VideoMatrixCoefficients::Bt709;
-  init.mPrimaries = VideoColorPrimaries::Bt709;
-  init.mTransfer = VideoTransferCharacteristics::Bt709;
-  decoderConfig.mColorSpace.Construct(init);
-
-  if (aRawData->mExtraData) {
-    auto& abov = decoderConfig.mDescription.Construct();
-    AutoEntryScript aes(aGlobal, "EncoderConfigToDecoderConfig");
-    size_t lengthBytes = aRawData->mExtraData->Length();
-    UniquePtr<uint8_t[], JS::FreePolicy> extradata(new uint8_t[lengthBytes]);
-    PodCopy(extradata.get(), aRawData->mExtraData->Elements(), lengthBytes);
-    JS::Rooted<JSObject*> description(
-        aes.cx(), JS::NewArrayBufferWithContents(aes.cx(), lengthBytes,
-                                                 std::move(extradata)));
-    JS::Rooted<JS::Value> value(aes.cx(), JS::ObjectValue(*description));
-    DebugOnly<bool> rv = abov.Init(aes.cx(), value);
-  }
-
-  return decoderConfig;
+  return VideoDecoderConfigInternal(
+      mOutputConfig.mCodec,        /* aCodec */
+      Some(mOutputConfig.mHeight), /* aCodedHeight */
+      Some(mOutputConfig.mWidth),  /* aCodedWidth */
+      Some(init),                  /* aColorSpace */
+      aRawData->mExtraData && !aRawData->mExtraData->IsEmpty()
+          ? Some(aRawData->mExtraData)
+          : Nothing(),                               /* aDescription*/
+      Maybe<uint32_t>(mOutputConfig.mDisplayHeight), /* aDisplayAspectHeight*/
+      Maybe<uint32_t>(mOutputConfig.mDisplayWidth),  /* aDisplayAspectWidth */
+      mOutputConfig.mHardwareAcceleration,           /* aHardwareAcceleration */
+      Nothing()                                      /*  aOptimizeForLatency */
+  );
 }
 
 #undef LOG